pkg/planner/core/rule_decorrelate.go

```diff
@@ -470,6 +514,133 @@ func (*DecorrelateSolver) Name() string {
+// It checks if the condition is of the form "outer_col = inner_col" where outer_col is a correlated column
+// from the Apply operator. Returns the inner column if it belongs to the given schema, otherwise returns nil.
+func extractJoinKeyFromCondition(apply *logicalop.LogicalApply, cond expression.Expression, schema *expression.Schema) *expression.Column {
+	decExpr := apply.DeCorColFromEqExpr(cond)
+	if decExpr == nil {
+		return nil
+	}
+	sf, ok := decExpr.(*expression.ScalarFunction)
+	if !ok || sf.FuncName.L != ast.EQ {
+		return nil
+	}
+	args := sf.GetArgs()
+	if len(args) != 2 {
+		return nil
+	}
+	innerCol, ok := args[1].(*expression.Column)
+	if !ok || !schema.Contains(innerCol) {
+		return nil
+	}
+	return innerCol
+}
+
+// isJoinKeyUniqueKey checks if join key is unique key.
+// Returns true if the join key forms a unique key constraint.
+func isJoinKeyUniqueKey(apply *logicalop.LogicalApply, plan base.LogicalPlan) bool {
+	var hasMultiRowOperator func(base.LogicalPlan) bool
+	hasMultiRowOperator = func(p base.LogicalPlan) bool {
+		// Use centralized function to check if operator can generate multiple rows
+		if logicalop.CanGenerateMultipleRows(p) {
+			return true
+		}
+		// Recursively check children
+		for _, child := range p.Children() {
+			if hasMultiRowOperator(child) {
+				return true
+			}
+		}
+		return false
+	}
+	if hasMultiRowOperator(plan) {
+		return false
+	}
+
+	// Extract join keys from Selection conditions and their children recursively
+	// Join conditions may be pushed down to DataSource or nested in child Selection nodes
+	innerJoinKeys := make([]*expression.Column, 0)
+
+	// Recursively extract all conditions from Selection nodes and their children
+	var extractConditions func(base.LogicalPlan)
+	extractConditions = func(p base.LogicalPlan) {
+		if sel, ok := p.(*logicalop.LogicalSelection); ok {
+			// Check conditions directly on Selection
+			for _, cond := range sel.Conditions {
+				if innerCol := extractJoinKeyFromCondition(apply, cond, sel.Schema()); innerCol != nil {
+					innerJoinKeys = append(innerJoinKeys, innerCol)
+				}
+			}
+			// Continue to check children recursively
+		} else if ds, ok := p.(*logicalop.DataSource); ok {
+			// Check conditions in DataSource (PushedDownConds may contain join key conditions)
+			for _, cond := range ds.PushedDownConds {
+				if innerCol := extractJoinKeyFromCondition(apply, cond, ds.Schema()); innerCol != nil {
+					innerJoinKeys = append(innerJoinKeys, innerCol)
+				}
+			}
+			// Stop recursion at DataSource
+			return
+		}
+		// Continue recursion for other nodes
+		for _, child := range p.Children() {
+			extractConditions(child)
+		}
+	}
+
+	extractConditions(plan)
+	if len(innerJoinKeys) == 0 {
+		return false
+	}
+
+	// Find the underlying DataSource to get PKOrUK
+	var findDataSource func(base.LogicalPlan) *logicalop.DataSource
+	findDataSource = func(p base.LogicalPlan) *logicalop.DataSource {
```

Comment:
Can we make it a unique function? Don't use an anonymous function.

Reply:
ok